### PR TITLE
Fixes in object detection image pipeline

### PR DESCRIPTION
**Pull request by eduramiba**
## What changes were proposed in this pull request?

 Fix IllegalArgumentException in ObjectDetectionRecordReader ImageObect creation after FlipImageTransform applied (ImageObect  does not allow flipped coordinates).

Fix wrong usage of random transforms in PipelineImageTransform::query (all transforms are queried even if they are not applied due to randomness).

Also add a few useful checks in ObjectDetectionRecordReader for correct bounds to help debug wrong image pipelines.

## How was this patch tested?

I modified these classes in my training code and used them to correctly train object detection with random image transforms.

### CODE REVIEW

**Comment by eduramiba** [comment]
Oh right, I will remove that check, but I guess we could still keep the < 0 check as that should never happen (this was the problem in my case, which made labels disappear of the training set after random flips but always being queried).

**Comment by AlexDBlack** [comment]
Yeah, I can see what you are going for there. I wonder if we could work out a way to check vs. the new image coords (i.e., translate 0/0 and w/h coords).

**Comment by eduramiba** [comment]
Actually, that made me realize that there is another bug in the querying, as original width and height are passed with the wrong assumption that the old width will map to the new width, and the same with height:

```java
float[] pts = imageTransform.query(io.getX1(), io.getY1(), io.getX2(), io.getY2(), (float)oW, (float)oH);
W = pts[4];
H = pts[5];
```

Then it's possible that it's still discarding some labels in my training silently, in the following section, as the image flip will map the original width to `-1`:

```java
if (cx < 0 || cx >= W || cy < 0 || cy >= H) {
    continue;
}
```

So yeah, I think we should obtain the new width and height in a different way and to the check against them, as you suggested. Maybe we should remove that continue anyway, since we will be checking for bounds, because discarding labels silently can lead to subtle wrong training without a clue that it's happening.

I will try to make add a fix in this pull request.

**Comment by eduramiba** [comment]
Hi,
I added a commit to fix that, and now all tests pass.

**Comment by eduramiba** [comment]
Actually wait... it's not correct if the image loader does a downscale directly instead of the transform. Let me try to fix it and add a test.

**Comment by eduramiba** [comment]
Ok, I think the latest commit is the correct way to do it.

**Comment by eduramiba** [comment]
It seems that querying a rotation transform can produce coordinates outside of the original image bound (no longer present in the transformed image), so maybe we should just keep the old `continue` condition (removing some labels if they go outside of the image) and never throw an exception.

**Comment by saudet** [comment]
If the transform pushes the center coordinates out of the image, the object should be removed, that's normal.

In any case, I guess we shouldn't try to transform the width and height as if they were coordinates, but creating a dummy image isn't the way either...

**Comment by eduramiba** [comment]
Agree, the dummy image is not great, but was the only solution I was able to come up with, without modifying other classes such as the image loader. I will try to find a better way that does not require many changes.

**Comment by eduramiba** [comment]
Maybe the `Image` class could have new fields containing dimensions after transform and after resize?

**Comment by eduramiba** [comment]
Hi,
I commited some changes to do it in a better way, without the need for a dummy image at all.

File: datavec-data/datavec-data-image/src/main/java/org/datavec/image/recordreader/objdetect/ObjectDetectionRecordReader.java [hunk 1]
```diff
@@ -184,22 +201,31 @@ public void initialize(InputSplit split) throws IOException {
...
         //put the label data into the output label array
         for (ImageObject io : objectsThisImg) {
             double cx = io.getXCenterPixels();
             double cy = io.getYCenterPixels();
-            double W = oW;
-            double H = oH;
             if (imageTransform != null) {
-                float[] pts = imageTransform.query(io.getX1(), io.getY1(), io.getX2(), io.getY2(), (float)oW, (float)oH);
-                io = new ImageObject(Math.round(pts[0]), Math.round(pts[1]), Math.round(pts[2]), Math.round(pts[3]), io.getLabel());
+                W = ((ProxyImageTransform) imageTransform).getImageWidthAfterTransform();
```
[the remark this answers is not in the document]
Comment (eduramiba) [reply datavec-data/datavec-data-image/src/main/java/org/datavec/image/recordreader/objdetect/ObjectDetectionRecordReader.java]: Hi,
Since the `imageTransform` is used by the `NativeImageLoader`, and it does the transform and rescaling in one step, it's not possible to always know the size of the image after being transformed (which may even include many transforms in a `PipelineImageTransform`).

But if we query the image transform to figure out where object coordinates will end up in the image, we must know the image dimensions after being transformed, not after being transformed and possibly rescaled by the image loader.

[the remark this answers is not in the document]
Comment (saudet) [reply datavec-data/datavec-data-image/src/main/java/org/datavec/image/recordreader/objdetect/ObjectDetectionRecordReader.java]: Yes, I see, you're right. Looks like a good solution. Could you move the functionality of ProxyImageTransform into BaseImageTransform though? Something like getCurrentImage() with maybe getHeight() and getWidth() implemented in ImageWritable. We'll have to make sure all the children set the new field, but it would make the functionality more reusable.

[the remark this answers is not in the document]
Comment (eduramiba) [reply datavec-data/datavec-data-image/src/main/java/org/datavec/image/recordreader/objdetect/ObjectDetectionRecordReader.java]: Sure, I will make these changes. I chose the proxy so not many classes needed changes, but if it's ok, we can do it that way.

[the remark this answers is not in the document]
Comment (eduramiba) [reply datavec-data/datavec-data-image/src/main/java/org/datavec/image/recordreader/objdetect/ObjectDetectionRecordReader.java]: Hi,
I implemented the requested changes, please review if they are what you asked.

**Comment by eduramiba** [comment]
Is It now ready for merge?
